src/lib.rs

```diff
@@ -1804,8 +1804,7 @@ impl WaitMode {
         // guaranteed to finish soon). Blocking waits should always join, even
         // in the presence of errors.
         match (self, expression_result) {
-            (&WaitMode::Blocking, _) => true,
-            (_, &Ok(Some(_))) => true,
+            (&WaitMode::Blocking, _) | (_, &Ok(Some(_))) => true,
             _ => false,
         }
     }
```

Comment:
Musing unrelated to this fix:

This while `should_join_background_thread` concept feels really complicated now that I'm getting back to it :( I've been planning to get rid of the `Then` concept entirely, because of the complications like this that it creates, and it might be that then I can delete this whole thing...